### PR TITLE
Delay the restart of application when a status report of failure is given

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

With the change in Fleet Server to report failure on error, this helps cleanup the flow to only restart if it stays failed for longer than 10 seconds. This allows a temporary failure to occur, before Elastic Agent would just force kill it and then restart it.

## Why is it important?

It is very possible that a user gets the connection information or authentication information to elasticsearch wrong when bootstrapping with Fleet Server. In that case the Elastic Agent should show a clear error message versus just spamming the logs with constant restarts.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- elastic/fleet-server#235

### CODE REVIEW

File: x-pack/elastic-agent/pkg/core/plugin/process/status.go [hunk 1]
```diff
@@ -35,21 +42,74 @@ func (a *Application) OnStatusChange(s *server.ApplicationState, status proto.St
...
-		if a.state.ProcessInfo != nil {
-			_ = a.state.ProcessInfo.Process.Kill()
-			a.state.ProcessInfo = nil
-		}
-		ctx := a.startContext
-		tag := a.tag
-
 		// it was marshalled to pass into the state, so unmarshall will always succeed
 		var cfg map[string]interface{}
 		_ = yaml.Unmarshal([]byte(s.Config()), &cfg)
```
Comment (ruflin) [inline x-pack/elastic-agent/pkg/core/plugin/process/status.go:47]: Unrelated to this PR but I don't think we should swallow the errors here.

File: x-pack/elastic-agent/pkg/core/plugin/process/status.go [hunk 2]
```diff
@@ -35,21 +42,74 @@ func (a *Application) OnStatusChange(s *server.ApplicationState, status proto.St
...
+	if a.state.ProcessInfo != nil {
+		_ = a.state.ProcessInfo.Process.Kill()
+		a.state.ProcessInfo = nil
+	}
+	ctx := a.startContext
+	tag := a.tag
+
+	err := a.start(ctx, tag, cfg)
+	if err != nil {
+		a.setState(state.Crashed, fmt.Sprintf("failed to restart: %s", err), nil)
```
Comment (ruflin) [inline x-pack/elastic-agent/pkg/core/plugin/process/status.go:113]: Can we add here a bit more info which process (name?) failed to restart?

**Review by ruflin — APPROVED** [review]
Code LGTM. What I wonder how to test this manually and in an automated way.

File: x-pack/elastic-agent/pkg/core/plugin/process/status.go (hunk 2 above)
Comment (blakerouse) [reply x-pack/elastic-agent/pkg/core/plugin/process/status.go:113]: That is there, that is managed inside of the `setState`.

File: x-pack/elastic-agent/pkg/core/plugin/process/status.go [hunk 3]
```diff
@@ -35,21 +37,74 @@ func (a *Application) OnStatusChange(s *server.ApplicationState, status proto.St
...
+	if a.restartCanceller != nil {
+		// already have running failed timer; just update config
+		a.restartConfig = cfg
+		return
+	}
+
+	ctx, cancel := context.WithCancel(a.startContext)
+	a.restartCanceller = cancel
+	a.restartConfig = cfg
+	t := time.NewTimer(a.processConfig.FailureTimeout)
```
Comment (scunningham) [inline x-pack/elastic-agent/pkg/core/plugin/process/status.go:64]: Should this start quick with exponential backoff to limit?

**Review by scunningham — APPROVED** [review]
Seems ok.  A bit hard to understand.  Is there a simpler way to write this?

File: x-pack/elastic-agent/pkg/core/plugin/process/status.go (hunk 3 above)
Comment (blakerouse) [reply x-pack/elastic-agent/pkg/core/plugin/process/status.go:64]: Feel like that would make it even more complicated, and harder to understand the time interval in log messages. A constant time allows for log messages to be clear that its every 10 seconds (or whatever setting value set) it is restarting.

**Comment by blakerouse** [comment]
> Seems ok. A bit hard to understand. Is there a simpler way to write this?

Based on how the application state interacts with the process applications in Elastic Agent, not really. Open to ideas.

**Pull request merged**